File: packages/core-dart/lib/src/routing/memo.dart
import '../address/codes.dart';

class NormalizeResult {
  final String? normalized;
  final List<Warning> warnings;

  NormalizeResult({this.normalized, required this.warnings});
}

final BigInt uint64Max = BigInt.parse('18446744073709551615');
final RegExp digitsOnly = RegExp(r'^\d+$');

NormalizeResult normalizeMemoTextId(String s) {
  final warnings = <Warning>[];

  // Step 1, 2, 3: Blank, non-digit
  if (s.isEmpty || !digitsOnly.hasMatch(s)) {
    return NormalizeResult(normalized: null, warnings: warnings);
  }

  // Step 4: Leading zeros
  var normalized = s.replaceFirst(RegExp(r'^0+'), '');
  if (normalized.isEmpty) {
    normalized = '0';
  }

  if (normalized != s) {
    warnings.add(
      Warning(
        code: WarningCode.nonCanonicalRoutingId,
        severity: 'warn',
        message:
            'Memo routing ID had leading zeros. Normalized to canonical decimal.',
        normalization: Normalization(original: s, normalized: normalized),
      ),
    );
  }

  // Step 5: uint64 max
  try {
    final val = BigInt.parse(normalized);
    if (val > uint64Max) {
      return NormalizeResult(normalized: null, warnings: warnings);
    }
  } catch (_) {
    return NormalizeResult(normalized: null, warnings: warnings);
  }

  return NormalizeResult(normalized: normalized, warnings: warnings);
}

File: packages/core-go/routing/memo.go
package routing

import (
	"math/big"
	"regexp"
	"strings"

	"github.com/stellar-address-kit/core-go/address"
)

var digitsOnly = regexp.MustCompile(`^\d+$`)
var uint64Max, _ = new(big.Int).SetString("18446744073709551615", 10)

type NormalizeResult struct {
	Normalized string
	Warnings   []address.Warning
}

func NormalizeMemoTextID(s string) NormalizeResult {
	warnings := []address.Warning{}

	// Step 1: Whitespace - not trimmed.
	// Step 2: Empty string.
	// Step 3: Any non-digit character.
	if s == "" || !digitsOnly.MatchString(s) {
		return NormalizeResult{Normalized: "", Warnings: warnings}
	}

	// Step 4: Strip leading zeros
	normalized := strings.TrimLeft(s, "0")
	if normalized == "" {
		normalized = "0"
	}

	if normalized != s {
		warnings = append(warnings, address.Warning{
			Code:     address.WarnNonCanonicalRoutingID,
			Severity: "warn",
			Message:  "Memo routing ID had leading zeros. Normalized to canonical decimal.",
			Normalization: &address.Normalization{
				Original:   s,
				Normalized: normalized,
			},
		})
	}

	// Step 5: Check uint64 max
	val, ok := new(big.Int).SetString(normalized, 10)
	if !ok || val.Cmp(uint64Max) > 0 {
		return NormalizeResult{Normalized: "", Warnings: warnings}
	}

	return NormalizeResult{Normalized: normalized, Warnings: warnings}
}
